### PR TITLE
[Rust]: Setup CI steps

### PR DESCRIPTION
**Pull request by satoshiotomakan**
## Description

As we want to write more rust, we will have to ensure the robustness of our implementations. to do this we need cargo caching, run the rust unit tests and display them on the CI, the clippy checks and finally the coverage check.

## Checklist

- [x] cargo caching (only for Linux Rust CI)
- [x] run rust unit tests
- [x] unit tests report in the CI
- [x] clippy check in the CI
- [x] coverage check + report CI

## TODO

- [ ] Consider installing only target required toolchains. For example, install `aarch64-apple-darwin` `x86_64-apple-darwin` for iOS CI only.
- [ ] Consider caching `rust/target` between each branch to improve compilation time. We can schedule a clean build every night to recompile the `rust/target` cache.

## Notes

Test coverage requires `cargo-llvm-cov` package (therefore `llvm-tools-preview` rustup tool) and nightly toolchain.
Test CI report requires `nextest` package.
We also use `udeps` `clippy` and `rustfmt` on nightly.

There should a possibility to cache the entire ~/.cargo directory to avoid recompilation each time.
For now, I use `sccache` to cache some crates.

### CODE REVIEW

**Comment by satoshiotomakan** [comment]
Test coverage requires `cargo-llvm-cov` package (therefore `llvm-tools-preview` rustup tool) and nightly toolchain.
Test CI report requires `nextest` package.
We also use `udeps` `clippy` and `rustfmt` on nightly.

It takes about 12 minutes to install all of the packages and tools, so I decided to cache the entire `~/./cargo` directory after the `tools/install-rust-dependencies` script.

File: tools/rust-lint [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+set -e
+
+pushd rust
+
+echo "Check code formatting"
+cargo fmt --check
+
+echo "Check Clippy warnings"
+cargo clippy -- -D warnings
```
Comment (Milerius) [inline tools/rust-lint:13]: where do we define the clippy warnings checks?

**Review by Milerius — APPROVED** [review]
Non-blocker comments, approved - I'm ok to have a separate CI for Rust, will facilitate the transition.

File: rust/coverage.stats [hunk 2]
```diff
@@ -0,0 +1 @@
+83.9
```
[the remark this answers is not in the document]
Comment (satoshiotomakan) [reply rust/coverage.stats:1]: Yes, of course we should aim probably more than 90%.
What do you think about adding the missing tests at the next PR?

File: tools/rust-lint (hunk 1 above)
Comment (satoshiotomakan) [reply tools/rust-lint:13]: We check the clippy warnings on the CI here
https://github.com/trustwallet/wallet-core/blob/826174f0c448f1de38646cdca0725be35da9c0ac/.github/workflows/linux-ci-rust.yml#L43-L45
Could you please clarify what you mean if this is not an answer to you question?

Comment (Milerius) [reply tools/rust-lint:13]: I mean how do you define specific warnings for clippy? is that just the default one?

File: rust/coverage.stats (hunk 2 above)
[the remark this answers is not in the document]
Comment (Milerius) [reply rust/coverage.stats:1]: yes

**Comment by hewigovens** [comment]
Looks good to me, can we cache for other pipelines as well?

**Comment by satoshiotomakan** [comment]
I haven't checked if the sccache may improve the performance significantly for other pipelines.
I would like to check it as a part of this suggestion https://github.com/trustwallet/wallet-core/issues/3044
> Looks good to me, can we cache for other pipelines as well?

**Pull request merged**